### PR TITLE
Update types for react-beautiful-dnd v12

### PR DESCRIPTION
**Pull request by aruniverse**
Please fill in this template.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x ] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/atlassian/react-beautiful-dnd/blob/master/src/types.js
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by PranavSenthilnathan** [comment]
Can you please take a look @varHarrie @bradleyayers @paustint @marknelissen @enricoboccadifuoco @lonyele @lukyth @aruniverse

**Comment by aruniverse** [comment]
@alexreardon maybe you want to review this?

**Comment by alexreardon** [comment]
I can take a look in the next 24hrs

File: types/react-beautiful-dnd/index.d.ts [hunk 1]
```diff
@@ -37,20 +41,11 @@ export interface ResponderProvided {
 
 export type OnBeforeDragStartResponder = (start: DragStart) => void;
 
-export type OnDragStartResponder = (
-    start: DragStart,
-    provided: ResponderProvided,
-) => void;
+export type OnDragStartResponder = (start: DragStart, provided: ResponderProvided) => void;
```
[the remark this answers is not in the document]
Comment (alexreardon) [reply types/react-beautiful-dnd/index.d.ts]: https://github.com/atlassian/react-beautiful-dnd/releases/tag/v12.1.0

File: types/react-beautiful-dnd/index.d.ts [hunk 2]
```diff
@@ -113,77 +178,94 @@ export interface DroppableStateSnapshot {
...
 
 export interface DraggableProvidedDragHandleProps {
+    onFocus: () => void;
+    onBlur: () => void;
     onMouseDown: React.MouseEventHandler<any>;
     onKeyDown: React.KeyboardEventHandler<any>;
     onTouchStart: React.TouchEventHandler<any>;
-    onTouchMove: React.TouchEventHandler<any>;
     'data-react-beautiful-dnd-drag-handle': string;
     'aria-roledescription': string;
```
Comment (alexreardon) [inline types/react-beautiful-dnd/index.d.ts]: this is now `aria-labelledby`

File: types/react-beautiful-dnd/index.d.ts [hunk 3]
```diff
@@ -113,77 +178,94 @@ export interface DroppableStateSnapshot {
...
 }
 
 export interface DraggableProvidedDragHandleProps {
+    onFocus: () => void;
+    onBlur: () => void;
     onMouseDown: React.MouseEventHandler<any>;
     onKeyDown: React.KeyboardEventHandler<any>;
     onTouchStart: React.TouchEventHandler<any>;
-    onTouchMove: React.TouchEventHandler<any>;
     'data-react-beautiful-dnd-drag-handle': string;
```
Comment (alexreardon) [inline types/react-beautiful-dnd/index.d.ts]:
```diff
type DragHandleProps = {|
+  // focus management is now handled without needing these handlers
-  onFocus: () => void,
-  onBlur: () => void,

+  // new sensor api does not require on element event handlers
-  onMouseDown: (event: MouseEvent) => void,
-  onKeyDown: (event: KeyboardEvent) => void,
-  onTouchStart: (event: TouchEvent) => void,

+  // Moving to new data-* format
-  'data-react-beautiful-dnd-drag-handle': string,
+  // what draggable the handle belongs to
+  'data-rbd-drag-handle-draggable-id': DraggableId,
+  // What DragDropContext the drag handle is in
+  'data-rbd-drag-handle-context-id': ContextId,

+  // New way of doing lift announcements
-  'aria-roledescription': string,
+  // id of drag handle aria description for screen readers
+  'aria-labelledby': ElementId,

  tabIndex: number,
  draggable: boolean,
  onDragStart: (event: DragEvent) => void,
|};
```

File: types/react-beautiful-dnd/index.d.ts [hunk 4]
```diff
@@ -83,7 +81,7 @@ export interface Combine {
 }
 
 export interface DropResult extends DragUpdate {
-    reason: DropReason;
+    reason: 'DROP' | 'CANCEL';
 }
 
 export interface DragDropContextProps {
```
Comment (alexreardon) [inline types/react-beautiful-dnd/index.d.ts]:
```diff
type Props = {|
  ...Responders,
  // We do not technically need any children for this component
  children: Node | null,
  // Read out by screen readers when focusing on a drag handle
+ liftInstruction?: string,
+ // Used for strict content security policies
+ // See our [content security policy guide](/docs/guides/content-security-policy.md)
+ nonce?: string,
+ // See our [sensor api](/docs/sensors/sensor-api.md)
+ sensors?: Sensor[],
+ enableDefaultSensors?: ?boolean,
|};
```

**Review by alexreardon — COMMENTED** [review]
Best things to do is check out:

- [12.0 release notes](https://github.com/atlassian/react-beautiful-dnd/releases/tag/v12.0.0)
- [Our manual type file](https://github.com/atlassian/react-beautiful-dnd/blob/master/docs/guides/types.md)

File: types/react-beautiful-dnd/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (aruniverse) [reply types/react-beautiful-dnd/index.d.ts]: Thanks, ill take a look and update accordingly. I made this when v12.0 was out, will need to update to support 12.1

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
Address feedback from @alexreardon

File: types/react-beautiful-dnd/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (aruniverse) [reply types/react-beautiful-dnd/index.d.ts]: I ended up copying from https://github.com/atlassian/react-beautiful-dnd/blob/master/src/types.js.

**Review by aruniverse — COMMENTED** [review]
addressed changes from @alexreardon

**Comment by aruniverse** [comment]
Can someone and review the changes ive made since @alexreardon initially commented?

@varHarrie @bradleyayers @paustint @marknelissen @enricoboccadifuoco @lonyele @lukyth

**Review by bradleyayers — APPROVED** [review]
This looks fairly benign, the old types are moved to v11 so compatibility should be maintained for anyone depending on 11.x, and so even if the new 12.x aren't perfect it should only affect people who explicitly upgrade. lgtm

**Review by aruniverse — COMMENTED** [review]
thank you @bradleyayers.

can you take a look @sheetalkamat @PranavSenthilnathan ?

**Review by aruniverse — COMMENTED** [review]
addressed changes and added definitions from rbd v11.0.4

**Comment by alexreardon** [comment]
Anything holding this one up?

**Comment by aruniverse** [comment]
We're waiting on @sheetalkamat to review

**Comment by ArielFrischer** [comment]
+1 @sheetalkamat would really appreciate a review, our team is stuck on old types.

**Review by bradleyayers — APPROVED** [review]

**Comment by leeyeh** [comment]
While waiting for the review, is there any preview/unstable/fork release available? Normally we can simply `npm install` a git brunch, but I dont know how it works for this monorepo.

**Comment by aruniverse** [comment]
I think a hack for now is to create a react-beautiful-dnd.d.ts locally in your project, and in it just declare the module which just exports all the v12 types in this pr. 

So:

'''
declare module "react-beautiful-dnd" {

// Just copy everything from the new type defs

}
'''

Tomorrow morning I'll try to create another PR from the same branch and see if we can get that approved instead by the community.

**Comment by aruniverse** [comment]
Created PR #42010 instead

**Comment by aruniverse** [comment]
Great news! [@types/react-beautiful-dnd@12.1.0](https://www.npmjs.com/package/@types/react-beautiful-dnd/v/12.1.0) has been published.

**Comment by silviuaavram** [comment]
I have one question about this line:
```tsx
getContainerForClone?: () => React.ReactElement<HTMLElement>;
```

In https://github.com/atlassian/react-beautiful-dnd/blob/master/docs/guides/reparenting.md#droppable---getcontainerforclone it clearly says that it wants an HTMLElement, and I implemented it successfully by providing something similar to `() => document.getElementById('whatever')`. However, types yell that it wants a React.Element.

What do you think? Should these types get updated to reflect this? Should the Readme in the original library be updated according to these types? And if yes, then what's the workaround suggested instead of `document.getElementById`.

Thanks!

**Pull request closed without merging**